### PR TITLE
Ensure diff doesn't persist patches

### LINKED ISSUE #852: Kubectl diff applies changes in 1.18.0
# What is the issue? 

When running kubectl diff in kubectl version 1.18.0 we are getting unexpected behavior of the diff being applied. What is expected is the output of diff (changes) without them being applied.  I went and verified on kubectl 1.16.x this was not the behavior of older clients. This has caused issues for us running checks on expected changes. 


Below is a simple test of Client and server on 1.18 on minikube. The test includes a basic nignx deployment with adding an ENV var.  

```bash
> kubectl version                                                                                                                                                                                                                                   
Client Version: version.Info{Major:"1", Minor:"18", GitVersion:"v1.18.0", GitCommit:"9e991415386e4cf155a24b1da15becaa390438d8", GitTreeState:"clean", BuildDate:"2020-03-26T06:17:09Z", GoVersion:"go1.14", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"18", GitVersion:"v1.18.0", GitCommit:"9e991415386e4cf155a24b1da15becaa390438d8", GitTreeState:"clean", BuildDate:"2020-03-25T14:50:46Z", GoVersion:"go1.13.8", Compiler:"gc", Platform:"linux/amd64"}
```

```bash
> kubectl get deployments my-nginx -o yaml |grep -i SHOULD_NOT_BE_HERE                                                                                                                                                                              
> vim test-dep.yml                                                                                                                                                                                                                                           
> kubectl diff -f test-dep.yml                                                                                                                                                                                                              
diff -u -N /var/folders/p0/4r8y9yds40q3fzyl2qv9lxp40000gp/T/LIVE-836964449/apps.v1.Deployment.default.my-nginx /var/folders/p0/4r8y9yds40q3fzyl2qv9lxp40000gp/T/MERGED-146007628/apps.v1.Deployment.default.my-nginx
--- /var/folders/p0/4r8y9yds40q3fzyl2qv9lxp40000gp/T/LIVE-836964449/apps.v1.Deployment.default.my-nginx	2020-04-02 12:18:56.000000000 -0700
+++ /var/folders/p0/4r8y9yds40q3fzyl2qv9lxp40000gp/T/MERGED-146007628/apps.v1.Deployment.default.my-nginx	2020-04-02 12:18:56.000000000 -0700
@@ -2,9 +2,9 @@
 kind: Deployment
 metadata:
   annotations:
-    deployment.kubernetes.io/revision: "2"
+    deployment.kubernetes.io/revision: "1"
   creationTimestamp: "2020-04-02T19:15:23Z"
-  generation: 4
+  generation: 5
   labels:
     app: my-nginx
   managedFields:
@@ -12,9 +12,7 @@
     fieldsType: FieldsV1
     fieldsV1:
       f:metadata:
-        f:annotations:
-          .: {}
-          f:deployment.kubernetes.io/revision: {}
+        f:annotations: {}
       f:status:
         f:availableReplicas: {}
         f:conditions:
@@ -33,17 +31,18 @@
             f:reason: {}
             f:status: {}
             f:type: {}
-        f:observedGeneration: {}
         f:readyReplicas: {}
         f:replicas: {}
         f:updatedReplicas: {}
     manager: kube-controller-manager
     operation: Update
-    time: "2020-04-02T19:17:40Z"
+    time: "2020-04-02T19:15:26Z"
   - apiVersion: apps/v1
     fieldsType: FieldsV1
     fieldsV1:
       f:metadata:
+        f:annotations:
+          f:deployment.kubernetes.io/revision: {}
         f:labels:
           .: {}
           f:app: {}
@@ -70,6 +69,11 @@
             f:containers:
               k:{"name":"nginx"}:
                 .: {}
+                f:env:
+                  k:{"name":"SHOULD_NOT_BE_HERE"}:
+                    .: {}
+                    f:name: {}
+                    f:value: {}
                 f:image: {}
                 f:imagePullPolicy: {}
                 f:name: {}
@@ -86,12 +90,13 @@
           k:{"type":"Progressing"}:
             f:lastUpdateTime: {}
             f:message: {}
+        f:observedGeneration: {}
     manager: kubectl
     operation: Update
-    time: "2020-04-02T19:17:40Z"
+    time: "2020-04-02T19:18:56Z"
   name: my-nginx
   namespace: default
-  resourceVersion: "4780"
+  resourceVersion: "4951"
   selfLink: /apis/apps/v1/namespaces/default/deployments/my-nginx
   uid: 857ffc2e-348d-4443-9992-e765ff1a77bf
 spec:
@@ -116,6 +121,8 @@
       - env:
         - name: DEMO_GREETING
           value: Hello from the environment
+        - name: SHOULD_NOT_BE_HERE
+          value: NOT GOOD!
         image: nginx
         imagePullPolicy: Always
         name: nginx
> kubectl get deployments my-nginx -o yaml |grep -i SHOULD_NOT_BE_HERE                                                                                                                                                                            
                  k:{"name":"SHOULD_NOT_BE_HERE"}:
        - name: SHOULD_NOT_BE_HERE
```

### PR DESCRIPTION
**Pull request by julianvmodesto**
**What type of PR is this?**

/kind bug

**What this PR does / why we need it**:
`kubectl diff` persists patches, which is bad since we should only be dry-running them.

**Which issue(s) this PR fixes**:

Fixes https://github.com/kubernetes/kubectl/issues/852

**Does this PR introduce a user-facing change?**:

```release-note
Fix kubectl diff so it doesn't actually persist patches
```

/assign @apelisse @seans3

### CODE REVIEW

File: staging/src/k8s.io/kubectl/pkg/cmd/apply/patcher.go [hunk 1]
```diff
@@ -185,7 +185,7 @@ func (p *Patcher) patchSimple(obj runtime.Object, modified []byte, source, names
 		}
 	}
 
-	patchedObj, err := p.Helper.Patch(namespace, name, patchType, patch, nil)
+	patchedObj, err := p.Helper.DryRun(p.ServerDryRun).Patch(namespace, name, patchType, patch, nil)
```
Comment (apelisse) [inline staging/src/k8s.io/kubectl/pkg/cmd/apply/patcher.go:188]: Can you also remove the code that sets dry-run on the helper in `newPatcher`?

File: test/cmd/diff.sh [hunk 2]
```diff
@@ -44,6 +44,14 @@ run_kubectl_diff_tests() {
     output_message=$(kubectl diff -f hack/testdata/pod-changed.yaml || test $? -eq 1)
     kube::test::if_has_string "${output_message}" 'k8s.gcr.io/pause:3.0'
 
+    # Ensure diff only dry-runs and doesn't persist change
```
Comment (apelisse) [inline test/cmd/diff.sh:47]: Can we make sure that both server-side and client-side dry-run are working properly? They both do things a different way so there is a possibility for a bug to creep-in one of the two.

**Comment by justaugustus** [comment]
Echoing [my sentiment from Slack](https://kubernetes.slack.com/archives/C2C40FMNF/p1585862747264900?thread_ts=1585862000.258700&cid=C2C40FMNF), this PR needs to also include a test to ensure we don't regress like this again.

/priority critical-urgent

/kind regression
(heh, @liggitt)

File: test/cmd/diff.sh (hunk 2 above)
Comment (liggitt) [reply test/cmd/diff.sh:47]: also, grabbing the resourceVersion before/after and ensuring it didn't change would be a more reliable way of making sure nothing was persisted (assuming nothing else is touching pods)

**Comment by apelisse** [comment]
Okay that looks good to me.
/lgtm

@liggitt, please have another look

**Comment by julianvmodesto** [comment]
Squashed!

**Comment by apelisse** [comment]
Still looks good
/lgtm

**Comment by apelisse** [comment]
Great, @julianvmodesto can you please create the cherry-pick for 1.18?

**Comment by sftim** [comment]
/kind regression
?

**Pull request merged**